### PR TITLE
Add support for webGL instancing

### LINKED ISSUE #6275: Instancing from WebGL2
### Increasing Access

This feature would allow for programs to run with better performance and for a lot more possibilities within the p5.js framework.

### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [X] WebGL
- [ ] Build Process
- [ ] Unit Testing
- [ ] Internalization
- [ ] Friendly Errors
- [ ] Other (specify if possible)

### Feature request details

Add a method that allows you to instance a certain amount of objects and let the user handle the rest through their own shader code.

### PR DESCRIPTION
**Pull request by RandomGamingDev**
Resolves #6275

 Changes:

This change adds instancing from WebGL2 to `endShape()` allowing for more performant rendering and more possibilities overall.

 Screenshots of the change:

Here's an example of instancing. The shapes below the rotated cube were all drawn within a single draw call.
Instancing

#### PR Checklist

- [x] `npm run lint` passes
- [x] [Inline documentation] is included / updated
- [x] [Unit tests] are included / updated

(I haven't had time to write a unit test yet nor am I sure what specific unit test to write)

This addition is based on a library that I wrote that works with the current latest release version of p5.js and that has its own demo.
https://github.com/RandomGamingDev/WebGLp5Instancing

[Inline documentation]: https://github.com/processing/p5.js/blob/main/contributor_docs/inline_documentation.md
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

### CODE REVIEW

**Comment by aferriss** [comment]
This is an interesting proposal, it would be good to see the accessibility reasoning behind this feature as well in the issue that you opened. I think performance is a valid argument but maybe there's something else as well. 

Personally, I think that it might make more sense to add the count argument into beginShape() instead of endshape(), but I'm open to being wrong about that if there's a good reason.

Can you also share some examples in the p5 web editor using the feature that we can play with? If this feature is accepted it will need examples anyway,  and it would be good to get an idea of how this looks in the shader, and what will be required of the user.

Should instanced geometry also be provided some instanceId attribute so that the user can manipulate instances within the shader? Furthermore, wouldn't this feature be more useful if there was an interface to create generic attribute buffers for instanced geometry? Otherwise, how is a user intended to translate / rotate / scale individual instances?

For unit tests you could think about testing that instanced geometry looks the same as non instanced, and that all the different geometry modes work correctly. Also test things like fill and stroke.

File: src/core/shape/vertex.js
Comment (aferriss) [inline src/core/shape/vertex.js:629]: I think it could be good to catch if the user tries to use a count < 1 and either log an error or silently fall back to just using a count of 1

File: src/webgl/p5.RendererGL.Immediate.js
Comment (aferriss) [inline src/webgl/p5.RendererGL.Immediate.js:395]: Why remove the .constants?

File: src/core/shape/vertex.js
Comment (aferriss) [inline src/core/shape/vertex.js:592]: Can this be renamed to "instanceCount" or something a little more descriptive?

File: src/webgl/p5.RendererGL.Immediate.js
Comment (RandomGamingDev) [reply src/webgl/p5.RendererGL.Immediate.js]: Sorry, that must've been a mistake from when I left them out for testing. I fixed it in the latest commit.

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js:629]: I was thinking of letting the WebGL implementation do the work since a count less than or equal to 0 could be valid, but prioritizing code safety does make more sense. I'll fix that in a new commit.

**Comment by RandomGamingDev** [comment]
I think that it'd work better with `endShape()`. I find it a bit off that beginShape is used to change the shape type and endShape to determine whether or not it closes, but understand that that's how you'd naturally think when you draw a shape. Following this logic and the logic of WebGL/OpenGL, it'd still make more sense to do the instancing at the end when everything's actually being drawn, and when you'd think about making duplicates a lot of the time.

This feature doesn't have to be limited to custom shapes, but since you oftentimes have to deal with vertices themselves, and need shader and WebGL knowledge to get it working, not to mention the fact that it could adversely effect p5.js in other unforseen ways. That's why I don't believe it'd make sense to add this feature to other shapes just to increase the complexity on the p5.js side and the confusion on the end user's side.

Shaders aren't applied to strokes, so I don't see why instancing, which is a shader-centric idea would be applied to strokes either, especially since it wouldn't have any effect unless the strokes were transparent, and even then have very little of an effect. I feel like the strokes, when using instancing would rather be more for debugging, especially since an experienced user of shaders would be able to add their own strokes in, strokes that would most likely fit better than p5.js strokes.

For the demo, I have a basic one in the library, although I do have one here that I used for debugging: https://editor.p5js.org/PotatoBoy/sketches/bC6WITRDd

The instance id attribute is handled by WebGL 2 using gl_InstanceID inside of the shader.

It would indeed be more useful with those, however it would raise the bar to use instancing, which is a very important feature in my opinion. I think that we should focus on making it more accessible. For those who need attributes, they can use constants, uniforms, and textures to store their data, or modify the p5.js library themselves until instancing gets added with attribute buffers.

As for the performance increase, create a bunch of objects and compare that to drawing them without instancing and the performance can be seen within chrome.

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js:592]: The problem I have with the name `instanceCount` is that it could be taken the wrong way. Perhaps something like `numInstances` would work, but I was thinking of keeping its name simple and understandable since drawing smth a certain amount of times makes sense in my view. If that's a problem then I'd probably change it to `numInstances`, but it might be less understandable rather than more.

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js:634]: It's been added to the front of the function below the parameter check in the latest commit.

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js:592]: If u look at the rest of the documentation most of the parameter names are decently simple.

**Comment by davepagurek** [comment]
> Personally, I think that it might make more sense to add the count argument into beginShape() instead of endshape(), but I'm open to being wrong about that if there's a good reason.

I don't have a strong opinion on this one, but maybe just think about what future extensions to this we might want to make (e.g. binding some per-instance attributes, or [having a per-instance transform matrix like Threejs does](https://threejs.org/docs/#api/en/objects/InstancedMesh.instanceMatrix)) and see if it would make it easier to do so in one location or the other (or if it makes no difference.) Any thoughts on what those might include @RandomGamingDev?

> Why would this feature need to be limited to custom shapes? Wouldn't it be more useful to instance things like lines or other basic shapes?

I think it might end up being easier to optimize performance for a feature like this if it's on a per shape level rather than something that happens inside of a primitive like a stroke. This way, we know for sure the user is intending to draw a bunch of the same thing, which should hopefully be a very clear optimization, whereas we've found it hard to get across-the-board performance improvements by adding instancing to all lines.

I do wonder, though, if we can do this in a way that works with `model()` too, as one of the things I'm going to be working on for the next release is a way to more easily make retained mode shapes, and pairing that with instancing could yield some great performance gains for particle system sorts of sketches.

> Shaders aren't applied to strokes, so I don't see why instancing, which is a shader-centric idea would be applied to strokes either

That's true for now while a shader + instance ID is the only way to make use of instanced rendering. If we think the most common use of instancing is to have a separate transformation matrix per instance, I can see us wanting to make a (non-shader) API specifically for this so that people don't have to write a shader to make use of it, and then we'd probably want the fills and strokes to both take that into account. That's probably the north star I'd want to work towards to balance usefulness and accessibility -- doesn't cover all cases, but would be much easier to write than making a custom shader, [especially while we don't have an easy way for users to augment our existing shaders without rewriting them.](https://github.com/processing/p5.js/issues/6144)

> As for the performance increase, create a bunch of objects and compare that to drawing them without instancing and the performance can be seen within chrome.

I think it would be good to have a minimum example of this both to use as a benchmark, and also to add to the p5 reference. As is, if a user adds a count to `endShape()`, all the instances would get drawn to the same spot and it would look like it's not working, so I think we need to provide a bit more documentation guidance on how the feature is intended to be used.

**Comment by RandomGamingDev** [comment]
I think those would include a system for adding attributes using `beginShape()`, `vertex()`, and `endShape()` since those r decently limited rn with the rest being handled by unifroms and textures.
Whether it's on a shape level or not shouldn't matter too much as long as the person using instancing knows enough to optimize for instancing. It would be possible to do it with `model()`, but it'd require modifying that function.
It'd be useful, but I'm not too sure what the reaction would be to modifying so much of the p5.js framework, especially for smth that a lot of people won't know how to use. If we want to go with the general API approach, we'd have to deal with things like position, size, rotation, scale, among other things. If we were to go through that route we'd most likely want to store that data in a texture that can then later be queried, but I'm not sure that that'd be too effective.
Alr, I'll add documentation later when I can.

**Comment by davepagurek** [comment]
> It would be possible to do it with model(), but it'd require modifying that function.

Makes sense, that could be a good feature to add to `model()`, but isn't required to be part of this PR.

> It'd be useful, but I'm not too sure what the reaction would be to modifying so much of the p5.js framework, especially for smth that a lot of people won't know how to use. If we want to go with the general API approach, we'd have to deal with things like position, size, rotation, scale, among other things. If we were to go through that route we'd most likely want to store that data in a texture that can then later be queried, but I'm not sure that that'd be too effective.

Solving for a more general case would maybe be better served by making an API for custom vertex attributes and allowing [the divisor](https://developer.mozilla.org/en-US/docs/Web/API/WebGL2RenderingContext/vertexAttribDivisor) to be changed so that it uses the same value for each instance. Looks like we could add that (and even add some nicer APIs for transformations specifically on top of that if we want) without running into issues here, which is good!

File: src/core/shape/vertex.js
Comment (davepagurek) [inline src/core/shape/vertex.js:562]: I think this is in the `endContour()` docs, did you want this example to be in `endShape()`?

**Comment by davepagurek** [comment]
Looks like there are some trailing spaces in the code causing the linter test to fail:
image

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js:562]: Yeah, sorry about that. I fixed it in the latest commit.

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js:682]: That was a mistake. I'll remove those.

**Comment by davepagurek** [comment]
@aferriss I'm thinking of merging this in as-is, knowing that for now it requires custom shaders to take advantage of instancing, and that this is only currently feasible for fills. Going forward, I'd like to improve upon both of those points either by:
1. making a way for users to augment our existing shaders without completely rewriting them (https://github.com/processing/p5.js/issues/6144) so that one could add some per-instance positioning logic
2. taking the threejs approach of making it easy to have [per-instance transforms](https://threejs.org/docs/#api/en/objects/InstancedMesh.instanceMatrix) and adding support for that into our default shaders. Any other properties requiring per-instance variation would still have to be done with a custom shader, which will work the same as it currently does, and might also benefit from (1).

Do any of those approaches sound ok to you? This current PR would be a step towards either of them.

**Comment by RandomGamingDev** [comment]
That sounds good.

File: src/core/shape/vertex.js
Comment (aferriss) [reply src/core/shape/vertex.js:592]: I was mainly thinking about the renaming in the context of helping future maintainers understand what this is for. I still think `count` is too ambiguous in this context. `numInstances` sounds fine to me if you prefer that.

File: src/core/shape/vertex.js
Comment (aferriss) [inline src/core/shape/vertex.js:630]: Can you add some comments to this shader to explain what's happening here? Maybe there's somewhere we can also add gl_InstanceID to the documentation?

File: src/core/shape/vertex.js
Comment (aferriss) [inline src/core/shape/vertex.js:646]: I think this is a stylistic choice, we can remove this comment.

File: src/core/shape/vertex.js
Comment (aferriss) [inline src/core/shape/vertex.js:691]: Are we actually checking for webGL2 mode anywhere? Wouldn't this just log if the user wasn't in webGL mode at all?

**Comment by aferriss** [comment]
@davepagurek That sounds fine to me. Before we merge though we still need to have a few more things:
- complete documentation for this feature, including: explanation of instancing, feature support (strokes not supported for ex), commented examples. I left more comments throughout the PR.
- unit tests
- error check for webGL2

**Comment by RandomGamingDev** [comment]
Yeah, I'll add more than just the example and also unit tests soon.
There's already a check for WebGL2 mode in `_drawImmediateFill` for if the instancing function fails since then we can be decently sure that they aren't using WebGL2.

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js:691]: This logs if the user isn't using WebGL mode at all, but there's already a check in place in `_drawImmediateFill()` which will trigger if trying to instance causes an error.

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js:646]: alr

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js:653]: alr I documented it in the top portion of text

**Comment by RandomGamingDev** [comment]
I added the documentation updates and I'll add the unit test when I get the time for it.

File: src/core/shape/vertex.js
Comment (aferriss) [inline src/core/shape/vertex.js:635]: can you simplify this line to be `vec4(float(gl_InstanceID) * 100.0, 0.0, 0.0, 0.0);`. There's no need for the vec2 here.

Comment (RandomGamingDev) [reply src/core/shape/vertex.js]: That looks good

**Comment by RandomGamingDev** [comment]
Oh forgot to update the thing. The unit tests were added.

File: src/core/shape/vertex.js
Comment (aferriss) [inline src/core/shape/vertex.js:597]: This is where we still need documentation explaining what instancing is, what it does, and when to use it. 

"When the `count` parameter is used with a value greater than 1, it enables instancing for shapes built when in WEBGL mode. Instancing is a feature that allows the GPU to efficiently draw multiples of the same shape. It's often used for particle effects or other times when you need a lot of repetition. In order to take advantage of instancing, you will also need to write your own custom shader using the `gl_InstanceID` keyword. You can read more about instancing [here](https://webglfundamentals.org/webgl/lessons/webgl-instanced-drawing.html) or by working from the example on this page."

Comment (RandomGamingDev) [reply src/core/shape/vertex.js:597]: Added the documentation u asked for.

**Review by aferriss — APPROVED** [review]
Thanks @RandomGamingDev I think it's looking good! Appreciate the patience :D

File: src/core/shape/vertex.js
Comment (davepagurek) [inline src/core/shape/vertex.js:663]: Not 100% sure if this is why this example is throwing an error compiling the shader, but it looks like the vertex shader uses a lowercase d in `Id` while the fragment shader capitalizes it

**Comment by RandomGamingDev** [comment]
I made it so that the shader compiled by changing the input and output names to be the same for `instanceID` and by using `flat` instead.

File: src/core/shape/vertex.js
Comment (RandomGamingDev) [reply src/core/shape/vertex.js]: That's on reason why, but int's also can't be interpolated like that so ur also supposed to use `flat`.

Comment (aferriss) [reply src/core/shape/vertex.js]: Ah sorry, that's my fault and what I get for editing the code inline in browser. Thanks for catching!

**Comment by davepagurek** [comment]
Great work on this @RandomGamingDev!

**Pull request merged**
